src/Npgsql/Replication/Logical/Protocol/LogicalReplicationProtocolMessage.cs

```diff
@@ -0,0 +1,37 @@
+﻿using JetBrains.Annotations;
+using NpgsqlTypes;
+using System;
+
+namespace Npgsql.Replication.Logical.Protocol
+{
+    /// <summary>
+    /// The base class of all Logical Replication Protocol Messages
+    /// </summary>
+    /// <remarks>
+    /// See https://www.postgresql.org/docs/current/protocol-logicalrep-message-formats.html for details about the
+    /// protocol.
+    /// </remarks>
+    [PublicAPI]
+    public abstract class LogicalReplicationProtocolMessage : INpgsqlReplicationMessage
```

Comment:
Yeah, I'm suffering here too.
Do you remember how hard it was for us to get the Logical Streaming Replication Protocol running because there's no documentation at all that you have to use the (integrated) pgutput logical decoding plugin.

I honestly don't know what would help our users more, if we'd talk about the Logical Streaming Replication Protocol all the time and name our classes after it (if you look at my namespace, you see that I integrated the name "Protocol" in it) because that's what you can find in the PostgreSQL documentation, or if we'd talk about pgoutput all the time which is ways terser if you want to name classes etc. but which is completely undocumented.
I opted for the middle (using a bit of both) trying to teach people to associate PgOutput with the Logical Streaming Replication Protocol.
I have no Idea how good or bad that is so I'm happy to take suggestions.

Reply:
> Npgsql prefix?

I'm suffering here.
It's in the namespace, disables IntelliSense for the first 6 characters and makes everything longer.

Reply:
>  Do you remember how hard it was for us to get the Logical Streaming Replication Protocol running because there's no documentation at all that you have to use the (integrated) pgutput logical decoding plugin.

Yeah, I remember - I've always been confused by this, and I agree that this confusion already exists in the PG docs themselves.

I'm also really not sure. It seems that originally, logical replication was very plugin-oriented (so pgoutput was just another option), but evolved to a point where it's the default (maybe even the only thing in use), and the plugins are maybe a bit irrelevant. If this is so, and pgoutput is simply the "default", then our API could also treat it as the default. For example, we could have user-facing CreateSlot and StartReplication methods which do pgoutput, and put TestEncoding in the name for test_encoding. In other words, I wouldn't even burden users with the need to know what "protocol "means - just use logical replication.

How does that sound?

Reply:
Re Npgsql prefix, I do agree we can probably drop it for types which are very specific and unlikely to appear elsewhere. NpgsqlConnection should be prefixed (because Connection is extremely generic and conflict-prone), but LogicalReplicationConnection is probably completely fine. In that case we should remove it from all the types under Npgsql.Replication.

Reply:
> but evolved to a point where it's the default (maybe even the only thing in use), and the plugins are maybe a bit irrelevant.

That's actually where the comment mentioned here https://github.com/npgsql/npgsql/pull/3281#discussion_r515472104 comes from.

If I had to write a plugin that serializes the PG WAL to JSON I'd probably rather consume the Logical Streaming Replication Protocol in Npgsql and write a JSON converter in C# than writing a PostgreSQL logical decoding plugin in C and pass it on in Npgsql, but that might only be me so I thought about plugins at two levels and tried to support them.

> This is also related to the namespace question - I'd consider just putting all this stuff directly under Npgsql.Replication.Logical (i.e. no need for a Protocol sub-namespace).

I Agree.

> LogicalReplicationConnection is probably completely fine. In that case we should remove it from all the types under Npgsql.Replication.

👍

Reply:
I thought about it some more and taken another look at the PG docs... you know the whole replication business much better than me - so I may be very wrong here. But from the impression I got, there's basically almost no reason for anyone to use anything other than pgoutput. It's definitely good that we have support for physical replication and non-pgoutput plugins, but even the PG docs seem to basically say that logical replication *is* pgoutput.

So my expectation would basically be to have everything needed for pgoutput be implicitly available under Npgsql.Replication (without requiring an extra namespace), and that once I have an NpgsqlLogicalReplicationConnection, I can just call CreateSlot and StartReplication to do pgoutput on it - as extensions or not, it doesn't really matter to the user. I understand that technically pgoutput is just another plugin, but the reality seems to be that it pretty much *is* logical replication.

Reply:
Removed the Npgsql prefix in https://github.com/Brar/npgsql/pull/9